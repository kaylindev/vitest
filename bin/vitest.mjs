#!/usr/bin/env node
'use strict'

import { fileURLToPath } from 'url'
import { resolve, dirname } from 'path'
import { startAndRun } from 'vite-node'
import minimist from 'minimist'

const argv = minimist(process.argv.slice(2), {
  string: ['root'],
})

const __dirname = dirname(fileURLToPath(import.meta.url))

await startAndRun({
  root: resolve(argv.root || process.cwd()),
  files: [
    resolve(__dirname, '../dist/cli.js'),
  ],
})
