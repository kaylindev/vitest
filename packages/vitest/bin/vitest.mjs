#!/usr/bin/env node

import { fileURLToPath } from 'url'
import { resolve } from 'pathe'
import { ensurePackageInstalled } from '../dist/utils.js'

const argv = process.argv.slice(2)

if (!await ensurePackageInstalled('vite'))
  process.exit(1)

if (argv.includes('--coverage')) {
  if (!await ensurePackageInstalled('c8'))
    process.exit(1)
  const filename = fileURLToPath(import.meta.url)
  const entry = resolve(filename, '../../dist/cli.js')
  process.argv.splice(2, 0, process.argv[0], entry)
  await import('c8/bin/c8.js')
}
else {
  await import('../dist/cli.js')
}
