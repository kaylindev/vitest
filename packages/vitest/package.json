{
  "name": "vitest",
  "version": "0.0.92",
  "description": "A blazing fast unit test framework powered by Vite",
  "keywords": [
    "vite",
    "vite-node",
    "test",
    "jest"
  ],
  "homepage": "https://github.com/antfu/vitest#readme",
  "bugs": {
    "url": "https://github.com/antfu/vitest/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/antfu/vitest.git"
  },
  "funding": "https://github.com/sponsors/antfu",
  "license": "MIT",
  "author": "Anthony Fu <anthonyfu117@hotmail.com>",
  "type": "module",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    },
    "./node": {
      "import": "./dist/node.js",
      "types": "./dist/node.d.ts"
    },
    "./global": {
      "types": "./global.d.ts"
    },
    "./*": "./*"
  },
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "bin": {
    "vitest": "./bin/vitest.mjs"
  },
  "files": [
    "dist",
    "bin",
    "*.d.ts"
  ],
  "scripts": {
    "build": "rimraf dist && rollup -c",
    "dev": "rollup -c --watch src",
    "typecheck": "tsc --noEmit",
    "prepublishOnly": "nr build"
  },
  "dependencies": {
    "@types/chai": "^4.3.0",
    "@types/chai-subset": "^1.3.3",
    "chai": "^4.3.4",
    "chai-subset": "^1.6.0",
    "fast-glob": "^3.2.7",
    "local-pkg": "^0.4.0",
    "piscina": "^3.2.0",
    "source-map": "^0.7.3",
    "tinyspy": "^0.1.0"
  },
  "devDependencies": {
    "@antfu/install-pkg": "^0.1.0",
    "@types/diff": "^5.0.1",
    "@types/jsdom": "^16.2.13",
    "@types/micromatch": "^4.0.2",
    "@types/natural-compare": "^1.4.1",
    "@types/node": "^17.0.0",
    "@types/prompts": "^2.0.14",
    "c8": "^7.10.0",
    "cac": "^6.7.12",
    "cli-truncate": "^3.1.0",
    "diff": "^5.0.0",
    "find-up": "^6.2.0",
    "flatted": "^3.2.4",
    "happy-dom": "^2.24.5",
    "jsdom": "^19.0.0",
    "log-update": "^5.0.0",
    "micromatch": "^4.0.4",
    "mlly": "^0.3.15",
    "nanoid": "^3.1.30",
    "natural-compare": "^1.4.0",
    "pathe": "^0.2.0",
    "picocolors": "^1.0.0",
    "pkg-types": "^0.3.2",
    "pretty-format": "^27.4.2",
    "prompts": "^2.4.2",
    "rollup": "^2.61.1",
    "strip-ansi": "^7.0.1",
    "typescript": "^4.5.4"
  },
  "peerDependencies": {
    "c8": "*",
    "happy-dom": "*",
    "jsdom": "*",
    "vite": "^2.7.1"
  },
  "peerDependenciesMeta": {
    "c8": {
      "optional": true
    },
    "happy-dom": {
      "optional": true
    },
    "jsdom": {
      "optional": true
    }
  },
  "engines": {
    "node": ">=16.0.0"
  }
}
