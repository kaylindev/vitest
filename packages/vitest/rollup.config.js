import fs from 'fs'
import path from 'pathe'
import esbuild from 'rollup-plugin-esbuild'
import dts from 'rollup-plugin-dts'
import resolve from '@rollup/plugin-node-resolve'
import commonjs from '@rollup/plugin-commonjs'
import json from '@rollup/plugin-json'
import alias from '@rollup/plugin-alias'
import license from 'rollup-plugin-license'
import c from 'picocolors'
import fg from 'fast-glob'
import pkg from './package.json'

const entry = [
  'src/index.ts',
  'src/node/cli.ts',
  'src/node.ts',
  'src/runtime/worker.ts',
  'src/runtime/entry.ts',
  'src/utils.ts',
]

const external = [
  ...Object.keys(pkg.dependencies),
  ...Object.keys(pkg.peerDependencies),
  'worker_threads',
]

export default ({ watch }) => [
  {
    input: entry,
    output: {
      dir: 'dist',
      format: 'esm',
    },
    external,
    plugins: [
      alias({
        entries: [
          { find: /^node:(.+)$/, replacement: '$1' },
        ],
      }),
      resolve({
        preferBuiltins: true,
      }),
      json(),
      commonjs(),
      esbuild({
        target: 'node14',
      }),
      !watch && licensePlugin(),
    ],
    onwarn(message) {
      if (/Circular dependencies/.test(message))
        return
      console.error(message)
    },
  },
  {
    input: [
      'src/index.ts',
    ],
    output: {
      file: 'dist/index.d.ts',
      format: 'esm',
    },
    external,
    plugins: [
      dts(),
    ],
  },
  {
    input: [
      'src/node.ts',
    ],
    output: {
      file: 'dist/node.d.ts',
      format: 'esm',
    },
    external,
    plugins: [
      dts(),
    ],
  },
]

function licensePlugin() {
  return license({
    thirdParty(dependencies) {
      // https://github.com/rollup/rollup/blob/master/build-plugins/generate-license-file.js
      // MIT Licensed https://github.com/rollup/rollup/blob/master/LICENSE-CORE.md
      const coreLicense = fs.readFileSync(
        path.resolve(__dirname, '../../LICENSE'),
      )
      function sortLicenses(licenses) {
        let withParenthesis = []
        let noParenthesis = []
        licenses.forEach((license) => {
          if (/^\(/.test(license))
            withParenthesis.push(license)

          else
            noParenthesis.push(license)
        })
        withParenthesis = withParenthesis.sort()
        noParenthesis = noParenthesis.sort()
        return [...noParenthesis, ...withParenthesis]
      }
      const licenses = new Set()
      const dependencyLicenseTexts = dependencies
        .sort(({ name: nameA }, { name: nameB }) =>
          nameA > nameB ? 1 : nameB > nameA ? -1 : 0,
        )
        .map(
          ({
            name,
            license,
            licenseText,
            author,
            maintainers,
            contributors,
            repository,
          }) => {
            let text = `## ${name}\n`
            if (license)
              text += `License: ${license}\n`

            const names = new Set()
            if (author && author.name)
              names.add(author.name)

            for (const person of maintainers.concat(contributors)) {
              if (person && person.name)
                names.add(person.name)
            }
            if (names.size > 0)
              text += `By: ${Array.from(names).join(', ')}\n`

            if (repository)
              text += `Repository: ${repository.url || repository}\n`

            if (!licenseText) {
              try {
                const pkgDir = path.dirname(
                  resolve(path.join(name, 'package.json'), {
                    preserveSymlinks: false,
                  }),
                )
                const licenseFile = fg.sync(`${pkgDir}/LICENSE*`, {
                  caseSensitiveMatch: false,
                })[0]
                if (licenseFile)
                  licenseText = fs.readFileSync(licenseFile, 'utf-8')
              }
              catch {}
            }
            if (licenseText) {
              text
                += `\n${
                  licenseText
                    .trim()
                    .replace(/(\r\n|\r)/gm, '\n')
                    .split('\n')
                    .map(line => `> ${line}`)
                    .join('\n')
                }\n`
            }
            licenses.add(license)
            return text
          },
        )
        .join('\n---------------------------------------\n\n')
      const licenseText
        = '# Vitest core license\n'
        + `Vitest is released under the MIT license:\n\n${
          coreLicense
        }\n# Licenses of bundled dependencies\n`
        + 'The published Vitest artifact additionally contains code with the following licenses:\n'
        + `${sortLicenses(licenses).join(', ')}\n\n`
        + `# Bundled dependencies:\n${
          dependencyLicenseTexts}`
      const existingLicenseText = fs.readFileSync('LICENSE.md', 'utf8')
      if (existingLicenseText !== licenseText) {
        fs.writeFileSync('LICENSE.md', licenseText)
        console.warn(
          c.yellow(
            '\nLICENSE.md updated. You should commit the updated file.\n',
          ),
        )
      }
    },
  })
}
