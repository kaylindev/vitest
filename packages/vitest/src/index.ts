import type { Anything, ArrayContaining } from './integrations/chai/jest-asymmetric-matchers'
import type { MatchersObject } from './integrations/chai/types'
import type { InlineConfig as VitestInlineConfig } from './types'

export { suite, test, describe, it } from './runtime/suite'
export * from './types'
export * from './runtime/hooks'
export * from './integrations/chai'
export * from './integrations/tinymock'
export * from './integrations/utils'

declare module 'vite' {
  interface UserConfig {
    /**
     * Options for Vitest
     */
    test?: VitestInlineConfig
  }
}

declare global {
  namespace Chai {
    interface ExpectStatic {
      extend(expects: MatchersObject): void
      stringContaining(expected: string): void
      anything(): Anything
      arrayContaining(expected: any): ArrayContaining
    }

    interface Assertion {
      // Chai
      chaiEqual(expected: any): void

      // Snapshot
      toMatchSnapshot(message?: string): Assertion
      toMatchInlineSnapshot(snapshot?: string, message?: string): Assertion
      matchSnapshot(message?: string): Assertion

      // Jest compact
      toEqual(expected: any): void
      toStrictEqual(expected: any): void
      toBe(expected: any): void
      toMatch(expected: string | RegExp): void
      toMatchObject(expected: any): void
      toContain(item: any): void
      toContainEqual(item: any): void
      toBeTruthy(): void
      toBeFalsy(): void
      toBeGreaterThan(num: number): void
      toBeGreaterThanOrEqual(num: number): void
      toBeLessThan(num: number): void
      toBeLessThanOrEqual(num: number): void
      toBeNaN(): void
      toBeUndefined(): void
      toBeNull(): void
      toBeDefined(): void
      toBeInstanceOf(c: any): void
      toBeCalledTimes(n: number): void
      toHaveLength(l: number): void
      toHaveProperty(p: string, value?: any): void
      toBeCloseTo(number: number, numDigits?: number): void
      toHaveBeenCalledTimes(n: number): void
      toHaveBeenCalledOnce(): void
      toHaveBeenCalled(): void
      toBeCalled(): void
      toHaveBeenCalledWith(...args: any[]): void
      toBeCalledWith(...args: any[]): void
      toHaveBeenNthCalledWith(n: number, ...args: any[]): void
      nthCalledWith(n: number, ...args: any[]): void
      toHaveBeenLastCalledWith(...args: any[]): void
      lastCalledWith(...args: any[]): void
      toThrow(expected?: string | RegExp): void
      toThrowError(expected?: string | RegExp): void
      toReturn(): void
      toHaveReturned(): void
      toReturnTimes(times: number): void
      toHaveReturnedTimes(times: number): void
      toReturnWith(value: any): void
      toHaveReturnedWith(value: any): void
      toHaveLastReturnedWith(value: any): void
      lastReturnedWith(value: any): void
      toHaveNthReturnedWith(nthCall: number, value: any): void
      nthReturnedWith(nthCall: number, value: any): void
    }
  }
}
