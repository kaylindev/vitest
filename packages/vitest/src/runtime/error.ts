import { format } from 'util'
import { stringify } from '../integrations/chai/jest-matcher-utils'

// https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm
export function serializeError(val: any): any {
  if (!val || typeof val === 'string')
    return val

  if (typeof val === 'function')
    return `Function<${val.name}>`
  if (typeof val !== 'object')
    return val
  if (val instanceof Promise || 'then' in val || (val.constructor && val.constructor.prototype === 'AsyncFunction'))
    return 'Promise'
  if (typeof Element !== 'undefined' && val instanceof Element)
    return val.tagName

  if (typeof val.asymmetricMatch === 'function')
    return `${val.toString()} ${format(val.sample)}`

  Object.keys(val).forEach((key) => {
    val[key] = serializeError(val[key])
  })

  return val
}

export function processError(err: any) {
  if (!err)
    return err
  // stack is not serialized in worker communication
  // we stringify it first
  if (err.stack)
    err.stackStr = String(err.stack)
  if (err.name)
    err.nameStr = String(err.name)

  if (typeof err.expected !== 'string')
    err.expected = stringify(err.expected)
  if (typeof err.actual !== 'string')
    err.actual = stringify(err.actual)

  return serializeError(err)
}
