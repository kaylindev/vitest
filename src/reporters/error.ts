import { promises as fs, existsSync } from 'fs'
import { relative } from 'path'
import c from 'picocolors'
import * as diff from 'diff'
import type { RawSourceMap } from 'source-map'
import { SourceMapConsumer } from 'source-map'
import cliTruncate from 'cli-truncate'
import { notNullish } from '../utils'
import type { VitestContext } from '../types'
import { F_POINTER } from './figures'

interface ErrorWithDiff extends Error {
  name: string
  nameStr?: string
  stack?: string
  stackStr?: string
  showDiff?: boolean
  actual?: any
  expected?: any
  operator?: string
}

interface Position {
  line: number
  column: number
}

declare global {
  namespace NodeJS {
    interface Process {
      __vitest__: VitestContext
    }
  }
}

export async function printError(error: unknown) {
  const ctx = process.__vitest__

  let e = error as ErrorWithDiff

  if (typeof error === 'string') {
    e = {
      message: error.split(/\n/g)[0],
      stack: error,
    } as any
  }

  const stackStr = e.stack || e.stackStr || ''
  const stacks = parseStack(stackStr)

  if (!stacks.length) {
    console.error(e)
  }
  else {
    const nearest = stacks.find((stack) => {
      return !stack.file.includes('vitest/dist')
      && ctx.server.moduleGraph.getModuleById(stack.file)
      && existsSync(stack.file)
    })

    printErrorMessage(e)
    await printStack(ctx, stacks, nearest, async(s, pos) => {
      if (s === nearest) {
        const sourceCode = await fs.readFile(nearest.file, 'utf-8')
        ctx.console.log(c.yellow(generateCodeFrame(sourceCode, 4, pos)))
      }
    })
  }

  if (e.showDiff)
    displayDiff(e.actual, e.expected)
}

async function getSourcePos(ctx: VitestContext, nearest: ParsedStack) {
  const mod = ctx.server.moduleGraph.getModuleById(nearest.file)
  const transformResult = mod?.ssrTransformResult
  const pos = await getOriginalPos(transformResult?.map, nearest)
  return pos
}

// TODO: handle big object and big string diff
function displayDiff(actual: string, expected: string) {
  console.error(c.gray(generateDiff(stringify(actual), stringify(expected))))
}

function printErrorMessage(error: ErrorWithDiff) {
  const errorName = error.name || error.nameStr || 'Unknown Error'
  console.error(c.red(`${c.bold(errorName)}: ${error.message}`))
}

async function printStack(
  ctx: VitestContext,
  stack: ParsedStack[],
  highlight?: ParsedStack,
  onStack?: ((stack: ParsedStack, pos: Position) => void),
) {
  if (!stack.length)
    return

  for (const frame of stack) {
    const pos = await getSourcePos(ctx, frame) || frame
    const color = frame === highlight ? c.yellow : c.gray
    const path = relative(ctx.config.root, frame.file)

    ctx.console.log(color(` ${c.dim(F_POINTER)} ${[frame.method, c.dim(`${path}:${pos.line}:${pos.column}`)].filter(Boolean).join(' ')}`))
    onStack?.(frame, pos)

    // reached at test file, skip the follow stack
    if (frame.file in ctx.state.filesMap)
      break
  }
  ctx.console.log()
}

function getOriginalPos(map: RawSourceMap | null | undefined, { line, column }: Position): Promise<Position | null> {
  return new Promise((resolve) => {
    if (!map)
      return resolve(null)

    SourceMapConsumer.with(map, null, (consumer) => {
      const pos = consumer.originalPositionFor({ line, column })
      if (pos.line != null && pos.column != null)
        resolve(pos as Position)
      else
        resolve(null)
    })
  })
}

const splitRE = /\r?\n/

export function posToNumber(
  source: string,
  pos: number | Position,
): number {
  if (typeof pos === 'number') return pos
  const lines = source.split(splitRE)
  const { line, column } = pos
  let start = 0
  for (let i = 0; i < line - 1; i++)
    start += lines[i].length + 1

  return start + column
}

export function numberToPos(
  source: string,
  offset: number | Position,
): Position {
  if (typeof offset !== 'number') return offset
  if (offset > source.length) {
    throw new Error(
      `offset is longer than source length! offset ${offset} > length ${source.length}`,
    )
  }
  const lines = source.split(splitRE)
  let counted = 0
  let line = 0
  let column = 0
  for (; line < lines.length; line++) {
    const lineLength = lines[line].length + 1
    if (counted + lineLength >= offset) {
      column = offset - counted + 1
      break
    }
    counted += lineLength
  }
  return { line: line + 1, column }
}

export function generateCodeFrame(
  source: string,
  indent = 0,
  start: number | Position = 0,
  end?: number,
  range = 2,
): string {
  start = posToNumber(source, start)
  end = end || start
  const lines = source.split(splitRE)
  let count = 0
  let res: string[] = []

  function lineNo(no: number | string = '') {
    return c.gray(`${String(no).padStart(3, ' ')}| `)
  }

  for (let i = 0; i < lines.length; i++) {
    count += lines[i].length + 1
    if (count >= start) {
      for (let j = i - range; j <= i + range || end > count; j++) {
        if (j < 0 || j >= lines.length)
          continue

        const lineLength = lines[j].length

        // to long, maybe it's a minified file, skip for codeframe
        if (lineLength > 200)
          return ''

        res.push(lineNo(j + 1) + cliTruncate(lines[j], process.stdout.columns - 5 - indent))

        if (j === i) {
          // push underline
          const pad = start - (count - lineLength)
          const length = Math.max(1, end > count ? lineLength - pad : end - start)
          res.push(lineNo() + ' '.repeat(pad) + c.red('^'.repeat(length)))
        }
        else if (j > i) {
          if (end > count) {
            const length = Math.max(1, Math.min(end - count, lineLength))
            res.push(lineNo() + c.red('^'.repeat(length)))
          }
          count += lineLength + 1
        }
      }
      break
    }
  }

  if (indent)
    res = res.map(line => ' '.repeat(indent) + line)

  return res.join('\n')
}

function stringify(obj: any) {
  // TODO: handle more types
  return String(obj)
}

const stackFnCallRE = /at (.*) \((.+):(\d+):(\d+)\)$/
const stackBarePathRE = /at ?(.*) (.+):(\d+):(\d+)$/

interface ParsedStack {
  method: string
  file: string
  line: number
  column: number
}

function parseStack(stack: string): ParsedStack[] {
  const lines = stack.split('\n')
  const stackFrames = lines.map((raw) => {
    const line = raw.trim()
    const match = line.match(stackFnCallRE) || line.match(stackBarePathRE)
    if (!match)
      return null

    let file = match[2]
    if (file.startsWith('file://'))
      file = file.slice(7)

    return {
      method: match[1],
      file: match[2],
      line: parseInt(match[3]),
      column: parseInt(match[4]),
    }
  })
  return stackFrames.filter(notNullish)
}

/**
 * Returns a diff between 2 strings with coloured ANSI output.
 *
 * @description
 * The diff will be either inline or unified dependent on the value
 * of `Base.inlineDiff`.
 *
 * @param {string} actual
 * @param {string} expected
 * @return {string} Diff
 */
export function generateDiff(actual: any, expected: any) {
  const diffSize = 2048
  if (actual.length > diffSize)
    actual = `${actual.substring(0, diffSize)} ... Lines skipped`

  if (expected.length > diffSize)
    expected = `${expected.substring(0, diffSize)} ... Lines skipped`

  return unifiedDiff(actual, expected)
}

/**
 * Returns unified diff between two strings with coloured ANSI output.
 *
 * @private
 * @param {String} actual
 * @param {String} expected
 * @return {string} The diff.
 */
function unifiedDiff(actual: any, expected: any) {
  const indent = '  '
  function cleanUp(line: string) {
    if (line[0] === '+')
      return indent + c.green(`${line[0]}${line.slice(1)}`)
    if (line[0] === '-')
      return indent + c.red(`${line[0]}${line.slice(1)}`)
    if (line.match(/@@/))
      return '--'
    if (line.match(/\\ No newline/))
      return null
    return indent + line
  }
  const msg = diff.createPatch('string', actual, expected)
  const lines = msg.split('\n').splice(5)
  return (
    `\n${indent}${c.red('- actual')}\n${indent}${c.green('+ expected')}\n\n${
      lines.map(cleanUp).filter(notBlank).join('\n')}`
  )
}

function notBlank(line: any) {
  return typeof line !== 'undefined' && line !== null
}
