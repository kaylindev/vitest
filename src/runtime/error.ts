export function processError(err: any) {
  if (!err)
    return err
  // stack is not serialized in worker communication
  // we stringify it first
  if (err.stack)
    err.stackStr = String(err.stack)
  if (err.name)
    err.nameStr = String(err.name)
  return err
}
