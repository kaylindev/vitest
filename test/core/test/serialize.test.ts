import { describe, expect, it } from 'vitest'
import { processError, serializeError } from '../../../packages/vitest/src/runtime/error'

describe('error serialize', () => {
  it('works', () => {
    expect(serializeError(undefined)).toEqual(undefined)
    expect(serializeError(null)).toEqual(null)
    expect(serializeError('hi')).toEqual('hi')

    expect(serializeError({
      foo: 'hi',
      promise: new Promise(() => {}),
      fn: () => {},
      null: null,
      nested: {
        false: false,
        class: class {},
      },
    })).toMatchSnapshot()
  })

  it('Should skip circular references to prevent hit the call stack limit', () => {
    const error: Record<string, any> = {
      toString: () => {
        return 'ops something went wrong'
      },
    }
    error.whatever = error
    error.whateverArray = [error, error]

    expect(serializeError(error)).toMatchSnapshot()
  })

  it('Should handle object with getter/setter correctly', () => {
    const user = {
      name: 'John',
      surname: 'Smith',

      get fullName() {
        return `${this.name} ${this.surname}`
      },
      set fullName(value) {
        [this.name, this.surname] = value.split(' ')
      },
    }

    expect(serializeError(user)).toEqual({
      name: 'John',
      surname: 'Smith',
      fullName: 'John Smith',
    })
  })
})

describe('Process Error', () => {
  it('Do not add expected/actual attributes in a object when any of both attributes does not exists', () => {
    const error = new Error('Ops something went wrong')

    const result = processError(error)

    expect(result.expected).not.toBeDefined()
    expect(result.actual).not.toBeDefined()
  })
})
