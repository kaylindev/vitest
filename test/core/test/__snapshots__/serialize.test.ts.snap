// Vitest Snapshot v1

exports[`error serialize > works 1`] = `
{
  "fn": "Function<fn>",
  "foo": "hi",
  "nested": {
    "class": "Function<class>",
    "false": false,
  },
  "null": null,
  "promise": "Promise",
}
`;
